From d482afec88b303303905958653fa906e65b2a1c2 Mon Sep 17 00:00:00 2001
From: Ankit Jain <radical@gmail.com>
Date: Fri, 18 Oct 2019 14:46:10 +0530
Subject: [PATCH] [mono] Linux build fixes (#128)

* [mono] Fix casing for `DotNet*` to be consistent and match the file on

.. disk.

* [mono] Remove Microsoft.DotNet.MSBuildSdkResolver on linux for the

.. bootstrap build. This is because it depends on a native lib,
libhostfxr*, which is not present in the bootrap msbuild, which breaks
the resolver and thus the build. Removing this should be fine, because
msbuild then just falls back to the bundled SDKs.

* [mono] Run the dotnet-install.sh script with `bash` instead of `sh` as

.. it breaks for `set -o pipefail` on ubuntu 16.04 .

* [mono] Skip copying the native lib, libhostfxr* for Linux

.. because this might or might not work on whatever distribution we are
actually building on.

* [mono] Sort both canonical and new files list, as the order can differ

.. between linux and macOS.
---
 eng/cibuild_bootstrapped_msbuild.sh                    | 5 +++++
 mono/build/extract_and_copy_hostfxr.sh                 | 2 +-
 mono/build/install.proj                                | 4 ++--
 mono/build/sdks_and_nugets/dotnet_resolver.proj        | 2 +-
 mono/build/sdks_and_nugets/update_sdks_and_nugets.proj | 9 ++++++---
 mono/build/update_bundled_bits.proj                    | 2 +-
 6 files changed, 16 insertions(+), 8 deletions(-)

diff --git a/eng/cibuild_bootstrapped_msbuild.sh b/eng/cibuild_bootstrapped_msbuild.sh
index 44d64ce58a..d94bfda7bd 100755
--- eng/cibuild_bootstrapped_msbuild.sh
+++ eng/cibuild_bootstrapped_msbuild.sh
@@ -62,6 +62,11 @@ function DownloadMSBuildForMono {
     # rename just to make it obvious when reading logs!
     mv $artifacts_dir/msbuild $mono_msbuild_dir
     chmod +x $artifacts_dir/mono-msbuild/MSBuild.dll
+
+    if [[ `uname -s` != 'Darwin' ]]; then
+        # with no .so available, this ends up breaking the build
+        rm -Rf $mono_msbuild_dir/SdkResolvers/Microsoft.DotNet.MSBuildSdkResolver
+    fi
     rm "$msbuild_zip"
   fi
 }
diff --git a/mono/build/extract_and_copy_hostfxr.sh b/mono/build/extract_and_copy_hostfxr.sh
index f4281bd900..d4b70026b7 100755
--- mono/build/extract_and_copy_hostfxr.sh
+++ mono/build/extract_and_copy_hostfxr.sh
@@ -44,7 +44,7 @@ OLDCWD=`pwd`
 cd $TMPDIR
 
 GetDotNetInstallScript $TMPDIR
-sh ./dotnet-install.sh --version $1 --install-dir $DOTNET_DIR --architecture x64 --runtime dotnet --skip-non-versioned-files
+bash ./dotnet-install.sh --version $1 --install-dir $DOTNET_DIR --architecture x64 --runtime dotnet --skip-non-versioned-files
 find $DOTNET_DIR -name libhostfxr.dylib | xargs -I {} cp -v {} $DESTDIR
 
 cd $OLDCWD
diff --git a/mono/build/install.proj b/mono/build/install.proj
index 93f5eeebb2..ed01b27d7c 100644
--- mono/build/install.proj
+++ mono/build/install.proj
@@ -186,11 +186,11 @@
         <Exec
             Condition="Exists('$(AllInstalledFiles_FileCanonical)')"
             IgnoreExitCode="$(IgnoreDiffFailure)"
-            Command="sort -o $(AllInstalledFiles_File) $(AllInstalledFiles_File) ; diff -u $(AllInstalledFiles_FileCanonical) $(AllInstalledFiles_File)" />
+            Command="sort -o $(AllInstalledFiles_FileCanonical) $(AllInstalledFiles_FileCanonical) ; sort -o $(AllInstalledFiles_File) $(AllInstalledFiles_File) ; diff -u $(AllInstalledFiles_FileCanonical) $(AllInstalledFiles_File)" />
 
         <Exec
             Condition="Exists('$(RemainingFiles_FileCanonical)')"
             IgnoreExitCode="$(IgnoreDiffFailure)"
-            Command="sort -o $(RemainingFiles_File) $(RemainingFiles_File); diff -u $(RemainingFiles_FileCanonical) $(RemainingFiles_File)" />
+            Command="sort -o $(RemainingFiles_FileCanonical) $(RemainingFiles_FileCanonical) ; sort -o $(RemainingFiles_File) $(RemainingFiles_File); diff -u $(RemainingFiles_FileCanonical) $(RemainingFiles_File)" />
     </Target>
 </Project>
diff --git a/mono/build/sdks_and_nugets/dotnet_resolver.proj b/mono/build/sdks_and_nugets/dotnet_resolver.proj
index 9606cac316..de689ce23d 100644
--- mono/build/sdks_and_nugets/dotnet_resolver.proj
+++ mono/build/sdks_and_nugets/dotnet_resolver.proj
@@ -8,7 +8,7 @@
 		<ItemGroup>
 			<FilesToCopy
                 Include="$(DependencyNuPkgPath)/lib/net472/**"
-                OutputDirectory="$(DotnetSdkResolverDir)" />
+                OutputDirectory="$(DotNetSdkResolverDir)" />
 
 		</ItemGroup>
     </Target>
diff --git a/mono/build/sdks_and_nugets/update_sdks_and_nugets.proj b/mono/build/sdks_and_nugets/update_sdks_and_nugets.proj
index c3670ea1d2..9d7db18173 100644
--- mono/build/sdks_and_nugets/update_sdks_and_nugets.proj
+++ mono/build/sdks_and_nugets/update_sdks_and_nugets.proj
@@ -1,6 +1,6 @@
 <Project DefaultTargets="DeploySdksAndNuGets">
     <Import Project="$(MSBuildThisFileDirectory)\..\common.props" />
-    <Import Project="$(MSBuildThisFileDirectory)\..\DotnetBitsVersions.props" />
+    <Import Project="$(MSBuildThisFileDirectory)\..\DotNetBitsVersions.props" />
 
     <PropertyGroup>
         <HostOSName Condition="'$(HostOSName)' == ''">osx</HostOSName>
@@ -49,7 +49,7 @@
                     DependencyPackageName=Microsoft.DotNet.MSBuildSdkResolver;
                     DependencyPackageVersion=$(MicrosoftDotNetMSBuildSdkResolverVersion);
                     NuGetPackagesDir=$(NuGetPackagesDir);
-                    DotnetSdkResolverDir=$(MSBuildSdkResolverOutDir)
+                    DotNetSdkResolverDir=$(MSBuildSdkResolverOutDir)
                 </Properties>
             </NuGetsToBundle>
         </ItemGroup>
@@ -61,6 +61,9 @@
             Targets="Restore;Build"
             Properties="OutputDirectory=$(DotNetOverlayDirectory)\nuget-support\msbuild-bin"/>
         
-        <Exec Command="$(MSBuildThisFileDirectory)/../extract_and_copy_hostfxr.sh $(DotNetSdkVersionForLibHostFxr) $(MSBuildSdkResolverOutDir)" />
+        <!-- Ignoring the .so for linux, because it works on a smaller set of distributions -->
+        <Exec
+            Condition="'$([MSBuild]::IsOSPlatform(OSX))'"
+            Command="$(MSBuildThisFileDirectory)/../extract_and_copy_hostfxr.sh $(DotNetSdkVersionForLibHostFxr) $(MSBuildSdkResolverOutDir)" />
     </Target>
 </Project>
diff --git a/mono/build/update_bundled_bits.proj b/mono/build/update_bundled_bits.proj
index eb4cee4d72..0f9ebd3a1e 100644
--- mono/build/update_bundled_bits.proj
+++ mono/build/update_bundled_bits.proj
@@ -1,5 +1,5 @@
 <Project DefaultTargets="FetchAndUpdateSdksAndNuGets">
-    <Import Project="$(MSBuildThisFileDirectory)\DotnetBitsVersions.props" />
+    <Import Project="$(MSBuildThisFileDirectory)\DotNetBitsVersions.props" />
 
     <Target Name="Build" DependsOnTargets="FetchAndUpdateSdksAndNuGets" />
 
